flit_core/flit_core/config.py

```diff
@@ -318,3 +368,250 @@ def _expand_requires_extra(re):
+
+    if 'dependencies' in proj:
+        _check_list_of_str(proj, 'dependencies')
+        md_dict['requires_dist'] = proj['dependencies']
+
+    if 'optional-dependencies' in proj:
+        _check_type(proj, 'optional-dependencies', dict)
+        optdeps = proj['optional-dependencies']
+        if not all(isinstance(e, list) for e in optdeps.values()):
+            raise ConfigError(
+                'Expected a dict of lists in optional-dependencies field'
+            )
+        for e, reqs in optdeps.items():
+            if not all(isinstance(a, str) for a in reqs):
+                raise ConfigError(
+                    'Expected a string list for optional-dependencies ({})'.format(e)
+                )
+
+        # Move dev-requires into requires-extra
+        reqs_noextra = md_dict.pop('requires_dist', [])
+        lc.reqs_by_extra = md_dict.pop('requires_extra', {})
+
+        # Add requires-extra requirements into requires_dist
+        md_dict['requires_dist'] = \
+            reqs_noextra + list(_expand_requires_extra(lc.reqs_by_extra))
+
+        md_dict['provides_extra'] = sorted(lc.reqs_by_extra.keys())
+
+        # For internal use, record the main requirements as a '.none' extra.
+        lc.reqs_by_extra['.none'] = reqs_noextra
+
+    if 'dynamic' in proj:
+        _check_list_of_str(proj, 'dynamic')
+        dynamic = set(proj['dynamic'])
+        unrec_dynamic = dynamic - {'version', 'description'}
+        if unrec_dynamic:
+            raise ConfigError(
+                "flit only supports dynamic metadata for 'version' & 'description'"
+            )
+        if dynamic.intersection(proj):
+            raise ConfigError(
+                "keys listed in project.dynamic must not be in [project] table"
+            )
+        lc.dynamic_metadata = dynamic
+
+    if ('version' not in proj) and ('version' not in lc.dynamic_metadata):
+        raise ConfigError(
+            "version must be specified under [project] or listed as a dynamic field"
+        )
+    if ('description' not in proj) and ('description' not in lc.dynamic_metadata):
+        raise ConfigError(
+            "description must be specified under [project] or listed as a dynamic field"
+        )
+
+    return lc
+
+def pep621_people(people, group_name='author') -> dict:
+    """Convert authors/maintainers from PEP 621 to core metadata fields"""
+    names, emails = [], []
+    for person in people:
+        if not isinstance(person, dict):
+            raise ConfigError("{} info must be list of dicts".format(group_name))
+        unrec_keys = set(person.keys()) - {'name', 'email'}
+        if unrec_keys:
+            raise ConfigError(
+                "Unrecognised keys in {} info: {}".format(group_name, unrec_keys)
+            )
+        if 'email' in person:
+            email = person['email']
+            if 'name' in person:
+                email = '{} <{}>'.format(person["name"], email)
```

Comment:
Users can include weird stuff in this field and break the formatting. Not sure if this is an issue.

```{ name = "Jane\n\n\nDoe <a@b.c>", email="jane.doe@example.org" }```

One *could* use [`formataddr`](https://docs.python.org/3/library/email.utils.html?highlight=formataddr#email.utils.formataddr).

Reply:
Thanks, I didn't know about that function. It somewhat garbles non-ASCII text, though:

```
In [10]: email.utils.formataddr(('Zoë', 'zoe@example.com>'))
Out[10]: '=?utf-8?q?Zo=C3=AB?= <zoe@example.com>>'
```

I don't know if that's what should happen. The original version metadata spec 1.0 ([PEP 241](https://www.python.org/dev/peps/pep-0241/)) says the format is based on email headers, and presumably that's what email headers do. But that was almost 20 years ago, and it's possible that tools expect something a bit more Unicode native now. PEP 621 explicitly suggests "the format `{name} <{email}>`".

I'll open a discussion on the Python discourse about this.

Reply:
https://discuss.python.org/t/core-metadata-email-fields-unicode/7421

Reply:
I assume most tools (tested with pip) handle UTF-8 just fine and don't need this garbling. I suggest instead to quote all names that contain problematic characters (e.g. `,<>"`) according to [RFC 5322](https://tools.ietf.org/html/rfc5322#section-3.2.3). Quoting is not necessary if the name just contains a non-ASCII character. Validating (internationalised) email-addresses is rather complicated and probably not worth it.

To ensure that the metadata format isn't corrupted one could check for each field that it doesn't contain linebreaks or control characters. The user probably knows that they shouldn't put newlines in their name, so an error message is sufficient.